src/components/Data/TypeLoader.ts

```diff
@@ -41,21 +46,108 @@ export class TypeLoader {
 		this.disposables = []
 	}
 
-	protected async load(typePath: string) {
-		// Check whether we have already loaded types
-		let src = types.get(typePath)
-		if (src) return src
+	protected async load(typeLocations: [string, string?][]) {
+		// Ignore if we are already loading types (e.g. if a tab has been switched while loading)
+		if (this.isLoading) return []
+		this.isLoading = true
 
-		await this.dataLoader.fired
+		const app = await App.getApp()
+
+		// Load the cache index which maps the urls to the location of the files in cache
+		let cacheIndex: Record<string, string> = {}
+		try {
+			cacheIndex = await app.fileSystem.readJSON(
+				`~local/data/cache/types/index.json`
+			)
+		} catch {}
+
+		// Create promises for loading each type definition. Once this resolves, the types will be cached appropriately
+		const toCache = await Promise.all(
+			typeLocations.map(([typeLocation, moduleName]) => {
+				return new Promise<[string, string, boolean, string?]>(
+					async (resolve) => {
+						// Before we try to load anything, make sure the type definition hasn't already been loaded and set in the type map
+						let src = types.get(typeLocation)
+						if (src) resolve([typeLocation, src, false, moduleName])
+
+						// Decide whether the type is being loaded from data or needs to be fetched externally
+						const isFromData = typeLocation.startsWith('types/')
+						if (isFromData) {
+							// Load types directly from bridge.'s data
+							await this.dataLoader.fired
+
+							const file = await this.dataLoader.readFile(
+								`data/packages/minecraftBedrock/${typeLocation}`
+							)
+							src = await file.text()
+							resolve([typeLocation, src, false, moduleName])
+							return
+						}
+
+						// First check cache to see if we have already cached the file, if so resolve with the file from cache
+						const cacheLocation = cacheIndex[typeLocation]
+						const file = cacheLocation
+							? await app.fileSystem
+									.readFile(cacheLocation)
+									.catch(() => null)
+							: null
+
+						// File is cached, so resolve with the file from cache
+						if (file) {
+							resolve([
+								typeLocation,
+								await file.text(),
+								false,
+								moduleName,
+							])
+							return
+						}
+
+						// The file couldn't be fetched from cache (because it is not in index or at the path specified)
+						// So we need to fetch it
+						const res = await fetch(typeLocation).catch(() => null)
+						// TODO: Maybe set a variable (failedToFetchAtLeastOnce) to later open an information window that tells the user that some types couldn't be fetched
```

Comment:
Fixed an error here when the user is not connected to a network. We can fix the TODO comment in a followup PR